benchmarks/graphsim.py

```diff
@@ -0,0 +1,168 @@
+"""
+Graph state simulator backends
+=======================================
+
+Here we benchmark our graph state simulator for MBQC with different backends.
+
+Currently, we have two backends: `networkx <https://networkx.org/documentation/stable/index.html>`_
+and `rustworkx <https://qiskit.org/ecosystem/rustworkx/index.html>`_.
+Both Python packages are used to manipulate graphs.
+While networkx is a pure Python package, rustworkx is a Rust package with Python bindings, which is faster than networkx.
+"""
+
+# %%
+# Firstly, let us import relevant modules:
+
+from copy import copy
+from time import perf_counter
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+from graphix import Circuit
+
+# %%
+# Next, define a function to generate random Clifford circuits.
+
+
+def genpair(n_qubits, count, rng):
+    pairs = []
+    for i in range(count):
+        choice = [j for j in range(n_qubits)]
+        x = rng.choice(choice)
+        choice.pop(x)
+        y = rng.choice(choice)
+        pairs.append((x, y))
+    return pairs
+
+
+def random_clifford_circuit(nqubits, depth, seed=42):
+    rng = np.random.default_rng(seed)
+    circuit = Circuit(nqubits)
+    gate_choice = list(range(5))
+    for _ in range(depth):
+        for j, k in genpair(nqubits, 2, rng):
+            circuit.cnot(j, k)
+        for j in range(nqubits):
+            k = rng.choice(gate_choice)
+            if k == 0:  # H
+                circuit.h(j)
+            elif k == 1:  # S
+                circuit.s(j)
+            elif k == 2:  # X
+                circuit.x(j)
+            elif k == 3:  # Z
+                circuit.z(j)
+            elif k == 4:  # Y
+                circuit.y(j)
+            else:
+                pass
+    return circuit
+
+
+# %%
+# We generate a set of random Clifford circuits with different widths.
+
+DEPTH = 3
+test_cases = [i for i in range(2, 300, 10)]
+graphix_patterns = {}
+
+for i in test_cases:
+    circuit = random_clifford_circuit(i, DEPTH)
+    pattern = circuit.transpile()
+    pattern.standardize()
+    nodes, edges = pattern.get_graph()
+    graphix_patterns[i] = (circuit, pattern, len(nodes))
```

Comment:
We recommend executing `pattern.shift_signals` here to fully reduce Pauli nodes.